client/consensus/manual-seal/src/seal_new_block.rs

```diff
@@ -0,0 +1,173 @@
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Block sealing utilities
+
+use crate::{Error, rpc};
+use std::sync::Arc;
+use sp_runtime::{
+	traits::{Block as BlockT, Header as HeaderT},
+	generic::BlockId,
+};
+use transaction_pool::txpool;
+use rpc::CreatedBlock;
+
+use consensus_common::{
+	self, BlockImport, Environment, Proposer,
+	ForkChoiceStrategy, BlockImportParams, BlockOrigin,
+	ImportResult, SelectChain,
+	import_queue::BoxBlockImport,
+};
+use sp_blockchain::HeaderBackend;
+use client_api::backend::Backend as ClientBackend;
+use hash_db::Hasher;
+use std::collections::HashMap;
+use std::time::Duration;
+use std::marker::PhantomData;
+
+/// params for sealing a new block
+pub struct SealBlockParams<'a, B: BlockT, C, CB, E, P: txpool::ChainApi, H> {
+	/// if true, empty blocks(without extrinsics) will be created.
+	/// otherwise, will return Error::EmptyTransactionPool.
+	pub create_empty: bool,
+	/// instantly finalize this block?
+	pub finalize: bool,
+	/// specify the parent hash of the about-to-created block
+	pub parent_hash: Option<<B as BlockT>::Hash>,
+	/// sender to report errors/success to the rpc.
+	pub sender: rpc::Sender<CreatedBlock<<B as BlockT>::Hash>>,
+	/// transaction pool
+	pub pool: Arc<txpool::Pool<P>>,
+	/// client backend
+	pub back_end: Arc<CB>,
+	/// Environment trait object for creating a proposer
+	pub env: &'a mut E,
+	/// SelectChain object
+	pub select_chain: &'a C,
+	/// block import object
+	pub block_import: &'a mut BoxBlockImport<B>,
+	/// inherent data provide
+	pub inherent_data_provider: &'a inherents::InherentDataProviders,
+	/// phantom data to pin the hasher type.
+	pub _phantom: PhantomData<H>
+}
+
+/// seals a new block with the given params
+pub async fn seal_new_block<B, C, CB, E, P, H>(params: SealBlockParams<'_, B, C, CB, E, P, H>)
+	where
+		B: BlockT,
+		H: Hasher<Out=<B as BlockT>::Hash>,
+		CB: ClientBackend<B, H>,
+		E: Environment<B>,
+		E::Error: std::fmt::Display,
+		<E::Proposer as Proposer<B>>::Error: std::fmt::Display,
+		P: txpool::ChainApi<Block=B, Hash=<B as BlockT>::Hash>,
+		C: SelectChain<B>,
+{
+	let SealBlockParams {
+		create_empty,
+		finalize,
+		pool,
+		parent_hash,
+		sender,
+		back_end,
+		select_chain,
+		block_import,
+		env,
+		inherent_data_provider,
+		..
+	} = params;
+
+	if pool.status().ready == 0 && !create_empty {
```

Comment:
This is subject to race conditions. The pool might not be empty now, but it will be empty when we start creating the block.
I think I'd go through the whole block production and then reject a block if it's actually empty, then risk unpredictable behavior here.

Reply:
> The pool might not be empty now, but it will be empty when we start creating the block.

Hmm, I don't see how this might be possible, can you elaborate?

Reply:
You can have multiple concurrent tasks running at `proposer.propose` since they are asynchronous. Also transaction pool culling happens asynchronously, so you can assume that the transaction you see in the pool right now are going to be there when you actually get pending iterator. The pool is designed to be accessed from multiple threads (or concurrent tasks), so there are race conditions.

I'd leave this check here as an optimisation, but then would double check if produced block has no transactions before importing.

Reply:
Also that got me thinking - what will happen if you trigger creation of two blocks at the same time (without providing `parent` parameter). I suspect that instead of producing 2 blocks on the canon chain it would produce two blocks at the same height, which most likely is not what we want.

Reply:
> what will happen if you trigger creation of two blocks at the same time (without providing parent parameter).

The authorship task processes `EngineCommands` synchronously not concurrently, so one will wait for the other.

>  but then would double check if produced block has no transactions before importing.

yeah this would ideal.